core/workspace_comment_svg.js

```diff
@@ -0,0 +1,591 @@
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.addSelect = function() {
+  Blockly.utils.addClass(
+      /** @type {!Element} */ (this.svgGroup_), 'blocklySelected');
+  this.setFocus();
+};
+
+/**
+ * Unselect this comment.  Remove its highlighting.
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.removeSelect = function() {
+  Blockly.utils.removeClass(
+      /** @type {!Element} */ (this.svgGroup_), 'blocklySelected');
+  this.blurFocus();
+};
+
+/**
+ * Focus this comment.  Highlight it visually.
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.addFocus = function() {
+  Blockly.utils.addClass(
+      /** @type {!Element} */ (this.svgGroup_), 'blocklyFocused');
+};
+
+/**
+ * Unfocus this comment.  Remove its highlighting.
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.removeFocus = function() {
+  Blockly.utils.removeClass(
+      /** @type {!Element} */ (this.svgGroup_), 'blocklyFocused');
+};
+
+/**
+ * Return the coordinates of the top-left corner of this comment relative to the
+ * drawing surface's origin (0,0), in workspace units.
+ * If the comment is on the workspace, (0, 0) is the origin of the workspace
+ * coordinate system.
+ * This does not change with workspace scale.
+ * @return {!goog.math.Coordinate} Object with .x and .y properties in
+ *     workspace coordinates.
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.getRelativeToSurfaceXY = function() {
+  var x = 0;
+  var y = 0;
+
+  var dragSurfaceGroup = this.useDragSurface_ ?
+      this.workspace.blockDragSurface_.getGroup() : null;
+
+  var element = this.getSvgRoot();
+  if (element) {
+    do {
+      // Loop through this comment and every parent.
+      var xy = Blockly.utils.getRelativeXY(element);
+      x += xy.x;
+      y += xy.y;
+      // If this element is the current element on the drag surface, include
+      // the translation of the drag surface itself.
+      if (this.useDragSurface_ &&
+          this.workspace.blockDragSurface_.getCurrentBlock() == element) {
+        var surfaceTranslation =
+            this.workspace.blockDragSurface_.getSurfaceTranslation();
+        x += surfaceTranslation.x;
+        y += surfaceTranslation.y;
+      }
+      element = element.parentNode;
+    } while (element && element != this.workspace.getBubbleCanvas() &&
+        element != dragSurfaceGroup);
+  }
+  this.xy_ = new goog.math.Coordinate(x, y);
+  return this.xy_;
+};
+
+/**
+ * Move a comment by a relative offset.
+ * @param {number} dx Horizontal offset, in workspace units.
+ * @param {number} dy Vertical offset, in workspace units.
+ * @package
+ */
+Blockly.WorkspaceCommentSvg.prototype.moveBy = function(dx, dy) {
+  var event = new Blockly.Events.CommentMove(this);
+  // TODO: Do I need to look up the relative to surface XY position here?
+  var xy = this.getRelativeToSurfaceXY();
+  this.translate(xy.x + dx, xy.y + dy);
+  this.xy_ = new goog.math.Coordinate(xy.x + dx, xy.y + dy);
```

Comment:
why not use the translate function of `goog.math.Coordinate`?

Reply:
also, `this.xy_` is getting set (to the same thing) twice because it's also getting set in translate...

Reply:
Can we reuse the code in the WorkspaceComment class?

Reply:
Removed extra setting of `this.xy_`.

I'd like to keep the updates of `this.xy_` and the actual transform attribute very tightly coupled.